Loop/Managers/LoopDataManager.swift

```diff
@@ -628,6 +630,55 @@ extension LoopDataManager {
             }
         }
     }
+    
+    /// Logs a new external bolus insulin dose in the DoseStore and HealthKit
+    ///
+    /// - Parameters:
+    ///   - startDate: The date the dose was started at.
+    ///   - value: The number of Units in the dose.
+    ///   - insulinModel: The type of insulin model that should be used for the dose.
+    func logOutsideInsulinDose(startDate: Date, units: Double, insulinModel: InsulinModel? = nil) {
+        let dose = DoseEntry(type: .bolus, startDate: startDate, value: units, unit: .units, insulinModel: insulinModel)
+        
+        logOutsideInsulinDose(dose: dose) { (error) in
+            if let error = error {
+                 self.logger.error(error)
+            } else {
+                self.recommendedBolus = nil
+                self.recommendedTempBasal = nil
+                self.insulinEffect = nil
+                self.notify(forChange: .bolus)
+            }
+        }
+    }
+    
+    /// Logs a new external bolus insulin dose in the DoseStore and HealthKit
+    ///
+    /// - Parameters:
+    ///   - dose: The dose to be added.
+    func logOutsideInsulinDose(dose: DoseEntry, completion: @escaping (_ error: DoseStore.DoseStoreError?) -> Void) {
+        let rawData = Data(UUID().uuidString.utf8)
+        
+        var events = [NewPumpEvent(date: dose.startDate, dose: dose, isMutable: false, raw: rawData, title: "External Bolus", type: .bolus)]
+        
+        if case .some(.tempBasal(let dose)) = basalDeliveryState {
+            // Ensure the currently-running temp basal is accounted for
+            // The raw data can be empty, as the event won't be persisted
+            let pendingBasal = NewPumpEvent(date: dose.startDate, dose: dose, isMutable: true, raw: Data(), title: String(describing: dose), type: .tempBasal)
+            events.append(pendingBasal)
+        }
+
+        if let model = dose.insulinModel {
+            doseStore.longestEffectDuration = max(doseStore.longestEffectDuration, model.effectDuration)
+        }
+
+        addPumpEvents(events, lastReconciliation: nil, updateLastReconciliation: false) { (error) in
```

Comment:
Ah, I see now. I don't know that we want to shoehorn a manual bolus in via pump event history.
It's worth creating a code path separate from `addPumpEvents`, factoring out any shared logic. That refactoring may require LoopKit changes as well, but should allow you to get rid of the `updateLastReconciliation` flag.